// @react-page/plugins-slate/lib/pluginFactories/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSimpleHtmlBlockPlugin = exports.createMarkPlugin = exports.createListPlugin = exports.createListItemPlugin = exports.createListIndentionPlugin = exports.createHeadingsPlugin = exports.createDataPlugin = exports.createComponentPlugin = void 0;
var createComponentPlugin_1 = __importDefault(require("./createComponentPlugin"));
exports.createComponentPlugin = createComponentPlugin_1.default;
var createDataPlugin_1 = __importDefault(require("./createDataPlugin"));
exports.createDataPlugin = createDataPlugin_1.default;
var createHeadingsPlugin_1 = __importDefault(require("./createHeadingsPlugin"));
exports.createHeadingsPlugin = createHeadingsPlugin_1.default;
var createListIndentionPlugin_1 = __importDefault(require("./createListIndentionPlugin"));
exports.createListIndentionPlugin = createListIndentionPlugin_1.default;
var createListItemPlugin_1 = __importDefault(require("./createListItemPlugin"));
exports.createListItemPlugin = createListItemPlugin_1.default;
var createListPlugin_1 = __importDefault(require("./createListPlugin"));
exports.createListPlugin = createListPlugin_1.default;
var createMarkPlugin_1 = __importDefault(require("./createMarkPlugin"));
exports.createMarkPlugin = createMarkPlugin_1.default;
var createSimpleHtmlBlockPlugin_1 = __importDefault(require("./createSimpleHtmlBlockPlugin"));
exports.createSimpleHtmlBlockPlugin = createSimpleHtmlBlockPlugin_1.default;
