// @react-page/editor/lib/core/service/hover/input.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeAndDispatchHover = exports.computeAndDispatchInsert = void 0;
var computeHover_1 = require("./computeHover");
var computeCurrentDropPosition = function (actions, hover, drag, monitor, element, cellPlugins) {
    var mousePosition = monitor.getClientOffset();
    var componentPosition = element.getBoundingClientRect();
    var room = {
        height: componentPosition.bottom - componentPosition.top,
        width: componentPosition.right - componentPosition.left,
    };
    var mouse = {
        y: mousePosition.y - componentPosition.top,
        x: mousePosition.x - componentPosition.left,
    };
    (0, computeHover_1.computeHover)(drag, hover, actions, {
        room: room,
        mouse: mouse,
        cellPlugins: cellPlugins,
    });
};
var computeAndDispatchInsert = function (hover, drag, monitor, element, actions, cellPlugins) {
    return computeCurrentDropPosition(actions, hover, drag, monitor, element, cellPlugins);
};
exports.computeAndDispatchInsert = computeAndDispatchInsert;
var computeAndDispatchHover = function (hover, drag, monitor, element, actions, cellPlugins) {
    return computeCurrentDropPosition(actions, hover, drag, monitor, element, cellPlugins);
};
exports.computeAndDispatchHover = computeAndDispatchHover;
