// @react-page/editor/lib/editor/StickyWrapper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var __1 = require("..");
var StickyWrapper = function (_a) {
    var children = _a.children;
    var ref = react_1.default.createRef();
    var sidebarPosition = (0, __1.useOption)('sidebarPosition');
    var stickyElRef = react_1.default.createRef();
    var _b = __read((0, react_1.useState)(false), 2), shouldStickToTop = _b[0], setShouldStickToTop = _b[1];
    var _c = __read((0, react_1.useState)(true), 2), shouldStickToBottom = _c[0], setShouldStickToBottom = _c[1];
    var _d = __read((0, react_1.useState)(0), 2), rightOffset = _d[0], setRightOffset = _d[1];
    var _e = __read((0, react_1.useState)(0), 2), rightOffsetFixed = _e[0], setRightOffsetFixed = _e[1];
    react_1.default.useEffect(function () {
        var calc = function () {
            var _a, _b, _c, _d, _e, _f, _g, _h;
            if (ref.current) {
                /**
                 * TODO: works, but needs refactor, could calculate style for sidebar directly
                 */
                var _j = ref.current.getBoundingClientRect(), top_1 = _j.top, left = _j.left, right = _j.right;
                var bottom = top_1 + ref.current.clientHeight;
                // document.documentElement.clientWidth is without scrollbars, so better for us
                if (sidebarPosition === 'rightAbsolute') {
                    var rightOffsetAbsolute = document.documentElement.clientWidth -
                        left -
                        ref.current.clientWidth;
                    setRightOffset(rightOffsetAbsolute);
                    setRightOffsetFixed(0);
                }
                else if (sidebarPosition === 'rightRelative') {
                    var rightOffsetAbsolute = document.documentElement.clientWidth -
                        left -
                        ref.current.clientWidth;
                    var rightOffsetRelative = (_a = stickyElRef.current) === null || _a === void 0 ? void 0 : _a.clientWidth;
                    var overlapsScreenRight = right + ((_b = stickyElRef.current) === null || _b === void 0 ? void 0 : _b.clientWidth) >
                        document.documentElement.clientWidth;
                    setRightOffset(overlapsScreenRight ? rightOffsetAbsolute : rightOffsetRelative);
                    setRightOffsetFixed(overlapsScreenRight
                        ? 0
                        : document.documentElement.clientWidth -
                            right -
                            ((_c = stickyElRef.current) === null || _c === void 0 ? void 0 : _c.clientWidth));
                }
                else if (sidebarPosition === 'leftRelative') {
                    var rightOffsetForRelativeLeft = -ref.current.clientWidth;
                    var rightOffsetForAbsoluteLeft = -ref.current.clientWidth -
                        left +
                        ((_e = (_d = stickyElRef.current) === null || _d === void 0 ? void 0 : _d.clientWidth) !== null && _e !== void 0 ? _e : 100);
                    var overlapsScreenLeft = right + rightOffsetForRelativeLeft <
                        stickyElRef.current.clientWidth;
                    setRightOffset(overlapsScreenLeft
                        ? rightOffsetForAbsoluteLeft
                        : rightOffsetForRelativeLeft);
                    setRightOffsetFixed(document.documentElement.clientWidth - left);
                }
                else if (sidebarPosition === 'leftAbsolute') {
                    console.warn('sidebarPosition leftAbsolute is currently buggy');
                    var rightOffsetForAbsoluteLeft = -ref.current.clientWidth -
                        left +
                        ((_g = (_f = stickyElRef.current) === null || _f === void 0 ? void 0 : _f.clientWidth) !== null && _g !== void 0 ? _g : 100);
                    setRightOffset(rightOffsetForAbsoluteLeft);
                    setRightOffsetFixed(document.documentElement.clientWidth -
                        ((_h = stickyElRef.current) === null || _h === void 0 ? void 0 : _h.clientWidth));
                }
                var uiHeight = stickyElRef.current
                    ? stickyElRef.current.clientHeight
                    : 400;
                setShouldStickToTop(top_1 > window.innerHeight - uiHeight);
                setShouldStickToBottom(bottom < window.innerHeight);
            }
        };
        document.addEventListener('scroll', calc);
        window.addEventListener('resize', calc);
        var observer = null;
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        if (global.IntersectionObserver) {
            observer = new IntersectionObserver(calc);
            if (ref.current) {
                observer.observe(ref.current);
            }
        }
        // do it once
        calc();
        return function () {
            document.removeEventListener('scroll', calc);
            window.removeEventListener('resize', calc);
            observer === null || observer === void 0 ? void 0 : observer.disconnect();
        };
    }, [ref, stickyElRef]);
    return (react_1.default.createElement("div", { style: { position: 'relative' }, ref: ref }, children({
        rightOffset: rightOffset,
        stickyElRef: stickyElRef,
        focusRef: ref,
        shouldStickToTop: shouldStickToTop,
        shouldStickToBottom: shouldStickToBottom,
        rightOffsetFixed: rightOffsetFixed,
    })));
};
exports.default = StickyWrapper;
