// @react-page/editor/lib/core/actions/cell/drag.js
"use strict";
/**
 * @module src/editor/actions/cell/drag
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.dragActions = exports.cancelCellDrag = exports.clearHover = exports.dragCell = exports.cellHoverInlineRight = exports.cellHoverInlineLeft = exports.cellHoverBelow = exports.cellHoverAbove = exports.cellHoverRightOf = exports.cellHoverLeftOf = exports.cellHover = exports.CLEAR_CLEAR_HOVER = exports.CELL_DRAG_CANCEL = exports.CELL_DRAG = exports.CELL_DRAG_HOVER = void 0;
var const_1 = require("../../const");
exports.CELL_DRAG_HOVER = 'CELL_DRAG_HOVER';
exports.CELL_DRAG = 'CELL_DRAG';
exports.CELL_DRAG_CANCEL = 'CELL_DRAG_CANCEL';
exports.CLEAR_CLEAR_HOVER = 'CLEAR_CLEAR_HOVER';
/**
 * Creates a redux action for when a cell hovers another item.
 *
 * @example
 * // const store = redux.createStore()
 * store.dispatch(cellHover(drag, hover, level, position))
 *
 * @param {PartialCell} drag The cell that is currently being dragged.
 * @param {Cell} hover The cell that is being hovered by the dragged cell.
 * @param {number} level Set the level if the dragged cells should hover over an ancestor of hover.
 * @param {string} position Can be left, right, above, below.
 * @return {Action}
 */
var cellHover = function (drag, hover, level, position) {
    var _a;
    if (level === void 0) { level = 0; }
    var hoverId = level === 0
        ? hover.id
        : (_a = hover.ancestorIds[Math.max(0, level - 1)]) !== null && _a !== void 0 ? _a : hover.id;
    return {
        type: exports.CELL_DRAG_HOVER,
        ts: new Date(),
        dragId: drag.id,
        hoverId: hoverId,
        level: level,
        position: position,
    };
};
exports.cellHover = cellHover;
/**
 * Creates a redux action for when a cell is hovering another cell on the left.
 *
 * @example
 * // const store = redux.createStore()
 * store.dispatch(cellHoverLeftOf(drag, hover, level))
 *
 * @param {PartialCell} drag The cell that is currently being dragged.
 * @param {Cell} hover The cell that is being hovered by the dragged cell.
 * @param {number} level Set the level if the dragged cells should hover over an ancestor of hover.
 * @return {Action}
 */
var cellHoverLeftOf = function (drag, hover, level) { return (0, exports.cellHover)(drag, hover, level, const_1.PositionEnum.LEFT_OF); };
exports.cellHoverLeftOf = cellHoverLeftOf;
/**
 * Creates a redux action for when a cell is hovering another cell on the right.
 *
 * @example
 * // const store = redux.createStore()
 * store.dispatch(cellHoverRightOf(drag, hover, level))
 *
 * @param {PartialCell} drag The cell that is currently being dragged.
 * @param {Cell} hover The cell that is being hovered by the dragged cell.
 * @param {number} level Set the level if the dragged cells should hover over an ancestor of hover.
 * @return {Action}
 */
var cellHoverRightOf = function (drag, hover, level) { return (0, exports.cellHover)(drag, hover, level, const_1.PositionEnum.RIGHT_OF); };
exports.cellHoverRightOf = cellHoverRightOf;
/**
 * Creates a redux action for when a cell is hovering another cell above.
 *
 * @example
 * // const store = redux.createStore()
 * store.dispatch(cellHoverAbove(drag, hover, level))
 *
 * @param {PartialCell} drag The cell that is currently being dragged.
 * @param {Cell} hover The cell that is being hovered by the dragged cell.
 * @param {number} level Set the level if the dragged cells should hover over an ancestor of hover.
 * @return {Action}
 */
var cellHoverAbove = function (drag, hover, level) { return (0, exports.cellHover)(drag, hover, level, const_1.PositionEnum.ABOVE); };
exports.cellHoverAbove = cellHoverAbove;
/**
 * Creates a redux action for when a cell is hovering another cell below.
 *
 * @example
 * // const store = redux.createStore()
 * store.dispatch(cellHoverBelow(drag, hover, level))
 *
 * @param {PartialCell} drag The cell that is currently being dragged.
 * @param {Cell} hover The cell that is being hovered by the dragged cell.
 * @param {number} level Set the level if the dragged cells should hover over an ancestor of hover.
 * @return {Action}
 */
var cellHoverBelow = function (drag, hover, level) { return (0, exports.cellHover)(drag, hover, level, const_1.PositionEnum.BELOW); };
exports.cellHoverBelow = cellHoverBelow;
/**
 * Creates a redux action for when a cell is hovering another cell on the left, but inline (css floating).
 *
 * @example
 * // const store = redux.createStore()
 * store.dispatch(cellHoverInlineLeft(drag, hover))
 *
 * @param {PartialCell} drag The cell that is currently being dragged.
 * @param {Cell} hover The cell that is being hovered by the dragged cell.
 * @return {Action}
 */
var cellHoverInlineLeft = function (drag, hover) {
    return (0, exports.cellHover)(drag, hover, 0, const_1.PositionEnum.INLINE_LEFT);
};
exports.cellHoverInlineLeft = cellHoverInlineLeft;
/**
 * Creates a redux action for when a cell is hovering another cell on the right, but inline (css floating).
 *
 * @example
 * // const store = redux.createStore()
 * store.dispatch(cellHoverInlineRight(drag, hover))
 *
 * @param {PartialCell} drag The cell that is currently being dragged.
 * @param {Cell} hover The cell that is being hovered by the dragged cell.
 * @return {Action}
 */
var cellHoverInlineRight = function (drag, hover) {
    return (0, exports.cellHover)(drag, hover, 0, const_1.PositionEnum.INLINE_RIGHT);
};
exports.cellHoverInlineRight = cellHoverInlineRight;
/**
 * Creates a redux action for when a cell is being dragged.
 *
 * @example
 * // const store = redux.createStore()
 * // const cell = { id: '1', ... }
 * store.dispatch(dragCell(cell.id))
 *
 * @param {string} id The id of the cell that is being dragged.
 * @return {Action}
 */
var dragCell = function (id) { return ({
    type: exports.CELL_DRAG,
    ts: new Date(),
    id: id,
}); };
exports.dragCell = dragCell;
/**
 * Creates a redux action to clear hover state of all cells.
 *
 * @return {Action}
 */
var clearHover = function () { return ({
    type: exports.CLEAR_CLEAR_HOVER,
    ts: new Date(),
}); };
exports.clearHover = clearHover;
/**
 * Creates a redux action for when cell dragging ends.
 *
 * @example
 * // const store = redux.createStore()
 * // const cell = { id: '1', ... }
 * store.dispatch(cancelCellDrag(cell.id))
 *
 * @param {string} id
 * @return {Action}
 */
var cancelCellDrag = function () { return ({
    type: exports.CELL_DRAG_CANCEL,
    ts: new Date(),
}); };
exports.cancelCellDrag = cancelCellDrag;
exports.dragActions = {
    cancelCellDrag: exports.cancelCellDrag,
    clearHover: exports.clearHover,
    dragCell: exports.dragCell,
    cellHoverInlineRight: exports.cellHoverInlineRight,
    cellHoverInlineLeft: exports.cellHoverInlineLeft,
    cellHoverBelow: exports.cellHoverBelow,
    cellHoverAbove: exports.cellHoverAbove,
    cellHoverRightOf: exports.cellHoverRightOf,
    cellHoverLeftOf: exports.cellHoverLeftOf,
    cellHover: exports.cellHover,
};
