// @react-page/editor/lib/core/utils/getCellStylingProps.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCellOuterDivClassName = exports.gridClass = exports.getCellInnerDivStylingProps = exports.getCellInnerDivClassName = exports.getCellInnerDivStyle = void 0;
var classnames_1 = __importDefault(require("classnames"));
var getCellInnerDivStyle = function (cell, plugin, data) {
    return (plugin === null || plugin === void 0 ? void 0 : plugin.cellStyle)
        ? typeof (plugin === null || plugin === void 0 ? void 0 : plugin.cellStyle) === 'function'
            ? plugin === null || plugin === void 0 ? void 0 : plugin.cellStyle(data)
            : plugin === null || plugin === void 0 ? void 0 : plugin.cellStyle
        : undefined;
};
exports.getCellInnerDivStyle = getCellInnerDivStyle;
var getCellInnerDivClassName = function (cell, plugin, data) {
    var _a;
    var additionalClass = (plugin === null || plugin === void 0 ? void 0 : plugin.cellClassName)
        ? typeof (plugin === null || plugin === void 0 ? void 0 : plugin.cellClassName) === 'function'
            ? plugin === null || plugin === void 0 ? void 0 : plugin.cellClassName(data)
            : plugin === null || plugin === void 0 ? void 0 : plugin.cellClassName
        : undefined;
    return ('react-page-cell-inner' +
        (((_a = cell.rows) === null || _a === void 0 ? void 0 : _a.length) > 0 ? '' : ' react-page-cell-inner-leaf') +
        (additionalClass ? ' ' + additionalClass : ''));
};
exports.getCellInnerDivClassName = getCellInnerDivClassName;
var getCellInnerDivStylingProps = function (cell, plugin, data) {
    return {
        style: (0, exports.getCellInnerDivStyle)(cell, plugin, data),
        className: (0, exports.getCellInnerDivClassName)(cell, plugin, data),
    };
};
exports.getCellInnerDivStylingProps = getCellInnerDivStylingProps;
var gridClass = function (size) {
    return "react-page-cell-sm-".concat(size || 12, " react-page-cell-xs-12");
};
exports.gridClass = gridClass;
var getCellOuterDivClassName = function (_a) {
    var _b;
    var size = _a.size, hasInlineNeighbour = _a.hasInlineNeighbour, inline = _a.inline, hasChildren = _a.hasChildren;
    return (0, classnames_1.default)('react-page-cell', (0, exports.gridClass)(size), (_b = {
            'react-page-cell-has-inline-neighbour': hasInlineNeighbour
        },
        _b["react-page-cell-inline-".concat(inline || '')] = inline,
        _b['react-page-cell-leaf'] = !hasChildren,
        _b));
};
exports.getCellOuterDivClassName = getCellOuterDivClassName;
