// @react-page/editor/lib/core/utils/removeUndefinedProps.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var removeUndefinedProps_1 = require("./removeUndefinedProps");
describe('removeUndefinedProps', function () {
    it('should remove undefined and null properties from object', function () {
        var obj = {
            a: 'a',
            b: undefined,
            c: 'something',
            d: null,
        };
        expect((0, removeUndefinedProps_1.removeUndefinedProps)(obj)).toEqual({
            a: 'a',
            c: 'something',
        });
    });
    it('does not touch nested stuff', function () {
        var obj = {
            a: 'a',
            b: undefined,
            c: 'something',
            d: {
                some: undefined,
                bar: 'bar',
            },
        };
        expect((0, removeUndefinedProps_1.removeUndefinedProps)(obj)).toEqual({
            a: 'a',
            c: 'something',
            d: {
                some: undefined,
                bar: 'bar',
            },
        });
    });
});
