// @react-page/editor/lib/core/actions/display.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setZoom = exports.resizeMode = exports.layoutMode = exports.previewMode = exports.editMode = exports.insertMode = exports.setMode = exports.setDisplayReferenceNodeId = exports.DEFAULT_DISPLAY_MODE = exports.DISPLAY_SET_ZOOM = exports.DISPLAY_MODE_RESIZING = exports.DISPLAY_MODE_INSERT = exports.DISPLAY_MODE_EDIT = exports.DISPLAY_MODE_LAYOUT = exports.DISPLAY_MODE_PREVIEW = exports.SET_PREVIOUS_DISPLAY_MODE = exports.SET_DISPLAY_MODE = exports.SET_DISPLAY_REFERENCE_NODE_ID = void 0;
exports.SET_DISPLAY_REFERENCE_NODE_ID = 'SET_DISPLAY_REFERENCE_NODE_ID';
exports.SET_DISPLAY_MODE = 'SET_DISPLAY_MODE';
exports.SET_PREVIOUS_DISPLAY_MODE = 'SET_PREVIOUS_DISPLAY_MODE';
exports.DISPLAY_MODE_PREVIEW = 'preview';
exports.DISPLAY_MODE_LAYOUT = 'layout';
exports.DISPLAY_MODE_EDIT = 'edit';
exports.DISPLAY_MODE_INSERT = 'insert';
exports.DISPLAY_MODE_RESIZING = 'resizing';
exports.DISPLAY_SET_ZOOM = 'DISPLAY_SET_ZOOM';
exports.DEFAULT_DISPLAY_MODE = exports.DISPLAY_MODE_EDIT;
var setDisplayMode = function (mode, referenceNodeId) {
    return function () { return ({
        type: exports.SET_DISPLAY_MODE,
        ts: new Date(),
        mode: mode,
        referenceNodeId: referenceNodeId,
    }); };
};
var setDisplayReferenceNodeId = function (referenceNodeId) { return ({
    type: exports.SET_DISPLAY_REFERENCE_NODE_ID,
    ts: new Date(),
    referenceNodeId: referenceNodeId,
}); };
exports.setDisplayReferenceNodeId = setDisplayReferenceNodeId;
/**
 * Dispatch to switch to arbitrary mode.
 */
var setMode = function (mode, referenceNodeId) { return ({
    type: exports.SET_DISPLAY_MODE,
    ts: new Date(),
    mode: mode,
    referenceNodeId: referenceNodeId,
}); };
exports.setMode = setMode;
/**
 * Dispatch to switch to insert display mode.
 */
exports.insertMode = setDisplayMode(exports.DISPLAY_MODE_INSERT);
/**
 * Dispatch to switch to edit display mode.
 */
exports.editMode = setDisplayMode(exports.DISPLAY_MODE_EDIT);
/**
 * Dispatch to switch to preview display mode.
 */
exports.previewMode = setDisplayMode(exports.DISPLAY_MODE_PREVIEW);
/**
 * Dispatch to switch to layout display mode.
 */
exports.layoutMode = setDisplayMode(exports.DISPLAY_MODE_LAYOUT);
/**
 * Dispatch to switch to resize display mode.
 */
exports.resizeMode = setDisplayMode(exports.DISPLAY_MODE_RESIZING);
/**
 * Dispatch Zoom the content
 */
var setZoom = function (zoom) { return ({
    type: exports.DISPLAY_SET_ZOOM,
    zoom: zoom,
}); };
exports.setZoom = setZoom;
