// @react-page/editor/lib/core/components/Row/ResizableRowCell.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var react_draggable_1 = __importDefault(require("react-draggable"));
var react_use_1 = require("react-use");
var Cell_1 = __importDefault(require("../Cell"));
var hooks_1 = require("../hooks");
var ResizableRowCell = function (_a) {
    var nodeId = _a.nodeId, rowWidth = _a.rowWidth, rowHasInlineChildrenPosition = _a.rowHasInlineChildrenPosition, isLast = _a.isLast, offset = _a.offset, size = _a.size, maxSize = _a.maxSize;
    var stepWidth = rowWidth / 12; // we're going to keep it a real number to preserve some precision
    var allowResizeInEditMode = (0, hooks_1.useOption)('allowResizeInEditMode');
    var isResizeMode = (0, hooks_1.useIsResizeMode)();
    var isEditMode = (0, hooks_1.useIsEditMode)();
    var isPreviewMode = (0, hooks_1.useIsPreviewMode)();
    var resize = (0, hooks_1.useResizeCell)(nodeId);
    var _b = __read((0, react_use_1.useMeasure)(), 2), ref = _b[0], cellHeight = _b[1].height;
    var cellSpacingY = (0, hooks_1.useCellSpacing)().y;
    var showResizeHandle = !isPreviewMode &&
        !isLast &&
        (isResizeMode || (allowResizeInEditMode && isEditMode));
    return (react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement(Cell_1.default, { nodeId: nodeId, measureRef: ref }),
        showResizeHandle ? (react_1.default.createElement(react_draggable_1.default, { bounds: {
                top: 0,
                bottom: 0,
                left: Math.round(stepWidth),
                right: Math.round(rowWidth - stepWidth),
            }, position: {
                x: rowHasInlineChildrenPosition === 'right'
                    ? Math.round(stepWidth * (12 - offset))
                    : Math.round(stepWidth * offset),
                y: 0,
            }, axis: "x", onDrag: function (e, data) {
                var diff = Math.round(data.deltaX / stepWidth);
                var newSize = rowHasInlineChildrenPosition === 'right'
                    ? size - diff
                    : size + diff;
                if (newSize > 0 && newSize <= maxSize)
                    resize(newSize);
            }, grid: [Math.round(stepWidth), 0] },
            react_1.default.createElement("div", { className: "resize-handle", style: {
                    // fix floating style
                    height: rowHasInlineChildrenPosition ? cellHeight : 'auto',
                    margin: cellSpacingY !== 0 ? "".concat(cellSpacingY / 2, "px 0") : undefined,
                }, onClick: function (e) { return e.stopPropagation(); } }))) : null));
};
exports.default = react_1.default.memo(ResizableRowCell);
