// @react-page/editor/lib/ui/AutoformControls/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoformControls = exports.AutoFields = exports.AutoField = exports.AutoForm = void 0;
var react_1 = __importStar(require("react"));
var hooks_1 = require("../../core/components/hooks");
var lazyLoad_1 = __importDefault(require("../../core/helper/lazyLoad"));
var makeUniformsSchema_1 = __importDefault(require("./makeUniformsSchema"));
exports.AutoForm = (0, lazyLoad_1.default)(function () { return Promise.resolve().then(function () { return __importStar(require('./AutoForm')); }); });
exports.AutoField = (0, lazyLoad_1.default)(function () { return Promise.resolve().then(function () { return __importStar(require('./AutoField')); }); });
exports.AutoFields = (0, lazyLoad_1.default)(function () { return Promise.resolve().then(function () { return __importStar(require('./AutoFields')); }); });
var getDefaultValue = function (bridge) {
    return bridge.getSubfields().reduce(function (acc, fieldName) {
        var _a;
        return (__assign(__assign({}, acc), (_a = {}, _a[fieldName] = bridge.getInitialValue(fieldName), _a)));
    }, {});
};
function AutoformControls(_a) {
    var onChange = _a.onChange, data = _a.data, schema = _a.schema, _b = _a.columnCount, columnCount = _b === void 0 ? 2 : _b, Content = _a.Content;
    var bridge = (0, react_1.useMemo)(function () { return (0, makeUniformsSchema_1.default)(schema); }, [schema]);
    (0, react_1.useEffect)(function () {
        var newDefaultData = __assign(__assign({}, getDefaultValue(bridge)), (data !== null && data !== void 0 ? data : {}));
        onChange(newDefaultData);
    }, [bridge]);
    var isSmall = (0, hooks_1.useIsSmallScreen)();
    return (react_1.default.createElement(exports.AutoForm, { model: data, autosave: true, schema: bridge, onSubmit: onChange }, Content ? (react_1.default.createElement(Content, { data: data, columnCount: columnCount })) : (react_1.default.createElement("div", { style: {
            columnCount: isSmall ? 1 : columnCount,
            columnRule: '1px solid #E0E0E0',
            columnGap: 48,
        } },
        react_1.default.createElement(exports.AutoFields, { element: react_1.Fragment })))));
}
exports.AutoformControls = AutoformControls;
