// @react-page/editor/lib/core/components/Cell/PluginControls/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@material-ui/core");
var styles_1 = require("@material-ui/styles");
var react_1 = __importStar(require("react"));
var ui_1 = require("../../../../ui");
var StyledTab = (0, styles_1.withStyles)(function () { return ({
    wrapper: {
        alignItems: 'flex-start',
    },
}); })(core_1.Tab);
var StyledTabs = (0, styles_1.withStyles)(function (theme) { return ({
    root: {
        marginTop: -12,
        marginBottom: -12,
        marginLeft: -24,
        alignItems: 'flex-start',
        backgroundColor: theme.palette.background.default,
    },
}); })(core_1.Tabs);
var ControlsList = react_1.default.memo(function (_a) {
    var _b;
    var controls = _a.controls, componentProps = _a.componentProps;
    var _c = __read((0, react_1.useState)(0), 2), tab = _c[0], setTab = _c[1];
    var activeControls = (_b = controls[tab]) === null || _b === void 0 ? void 0 : _b.controls;
    return (react_1.default.createElement("div", { style: {
            display: 'flex',
            flexDirection: 'row',
        } },
        react_1.default.createElement(StyledTabs, { value: tab, onChange: function (e, v) { return setTab(v); }, orientation: "vertical", variant: "scrollable" }, controls.map(function (t, index) { return (react_1.default.createElement(StyledTab, { label: t.title, key: index })); })),
        activeControls ? (react_1.default.createElement("div", { style: {
                flex: 1,
                marginLeft: 24,
                display: 'flex',
            } },
            react_1.default.createElement(Controls, { controls: activeControls, componentProps: componentProps }))) : null));
});
var Controls = react_1.default.memo(function (_a) {
    var controls = _a.controls, componentProps = _a.componentProps;
    var pluginControls = null;
    if (Array.isArray(controls)) {
        return react_1.default.createElement(ControlsList, { componentProps: componentProps, controls: controls });
    }
    if ((controls === null || controls === void 0 ? void 0 : controls.type) === 'custom') {
        var Component = controls.Component;
        pluginControls = react_1.default.createElement(Component, __assign({}, componentProps, controls));
    }
    else if ((controls === null || controls === void 0 ? void 0 : controls.type) === 'autoform') {
        pluginControls = react_1.default.createElement(ui_1.AutoformControls, __assign({}, componentProps, controls));
    }
    return react_1.default.createElement("div", { style: { overflow: 'auto', flex: 1 } }, pluginControls);
});
var PluginControls = function (_a) {
    var controls = _a.controls, componentProps = _a.componentProps;
    return (react_1.default.createElement("div", { style: {
            maxHeight: '50vh',
            // if it has tabs, stretch to avoid jumping tabs
            width: Array.isArray(controls) ? '100vw' : undefined,
            maxWidth: '100%',
            display: 'flex',
        } },
        react_1.default.createElement(Controls, { controls: controls, componentProps: componentProps })));
};
exports.default = react_1.default.memo(PluginControls);
