// @react-page/editor/lib/renderer/HTMLRenderer.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HTMLRenderer = void 0;
var classnames_1 = __importDefault(require("classnames"));
var react_1 = __importDefault(require("react"));
var NoopProvider_1 = __importDefault(require("../core/components/Cell/NoopProvider"));
var migrate_1 = require("../core/migrations/migrate");
var optimize_1 = require("../core/reducer/value/helper/optimize");
var setAllSizesAndOptimize_1 = require("../core/reducer/value/helper/setAllSizesAndOptimize");
var getAvailablePlugins_1 = require("../core/utils/getAvailablePlugins");
var getCellData_1 = require("../core/utils/getCellData");
var getCellSpacing_1 = require("../core/utils/getCellSpacing");
var getCellStylingProps_1 = require("../core/utils/getCellStylingProps");
var rowHasInlineChildren = function (_a) {
    var cells = _a.cells;
    return Boolean(cells.length === 2 && Boolean(cells[0].inline));
};
var HTMLRow = react_1.default.memo(function (_a) {
    var _b = _a.cells, cells = _b === void 0 ? [] : _b, className = _a.className, lang = _a.lang, cellPlugins = _a.cellPlugins, cellSpacing = _a.cellSpacing;
    return (react_1.default.createElement("div", { className: (0, classnames_1.default)('react-page-row', className, {
            'react-page-row-has-floating-children': rowHasInlineChildren({ cells: cells }),
        }), style: { margin: "0 ".concat(-cellSpacing.x / 2, "px") } }, cells.map(function (c) { return (react_1.default.createElement(HTMLCell, __assign({ key: c.id }, c, { lang: lang, cellPlugins: cellPlugins, cellSpacing: cellSpacing }))); })));
});
// eslint-disable-next-line no-empty-function
var noop = function () {
    return;
};
var HTMLCell = react_1.default.memo(function (props) {
    var _a, _b, _c, _d, _e, _f;
    var lang = props.lang, cellPlugins = props.cellPlugins, cellSpacing = props.cellSpacing, cell = __rest(props, ["lang", "cellPlugins", "cellSpacing"]);
    var size = cell.size, hasInlineNeighbour = cell.hasInlineNeighbour, inline = cell.inline, isDraftI18n = cell.isDraftI18n, isDraft = cell.isDraft;
    var hasChildren = ((_a = cell.rows) === null || _a === void 0 ? void 0 : _a.length) > 0;
    if ((_b = isDraftI18n === null || isDraftI18n === void 0 ? void 0 : isDraftI18n[lang]) !== null && _b !== void 0 ? _b : isDraft) {
        return null;
    }
    var data = (_c = (0, getCellData_1.getCellData)(cell, lang)) !== null && _c !== void 0 ? _c : {};
    var plugin = cell.plugin
        ? cellPlugins.find(function (p) { return p.id === cell.plugin.id; })
        : null;
    var outerClasses = (0, getCellStylingProps_1.getCellOuterDivClassName)({
        hasChildren: hasChildren,
        size: size,
        hasInlineNeighbour: hasInlineNeighbour,
        inline: inline,
    });
    if (plugin) {
        var Renderer = plugin.Renderer;
        var Provider = plugin.Provider && !plugin.disableProviderInReadOnly
            ? plugin.Provider
            : NoopProvider_1.default;
        var pluginCellSpacing_1 = (0, getCellSpacing_1.getPluginCellSpacing)(plugin, data);
        if (typeof pluginCellSpacing_1 === 'undefined' || pluginCellSpacing_1 == null) {
            pluginCellSpacing_1 = cellSpacing;
        }
        else {
            pluginCellSpacing_1 = (0, getCellSpacing_1.normalizeCellSpacing)(pluginCellSpacing_1);
        }
        var props_1 = {
            readOnly: true,
            lang: lang,
            nodeId: cell.id,
            data: data,
            onChange: noop,
            pluginConfig: plugin,
            focused: false,
            isPreviewMode: false,
            isEditMode: false,
        };
        var childCellPlugins_1 = (0, getAvailablePlugins_1.getChildCellPlugins)(cellPlugins, {
            data: data,
            pluginId: plugin === null || plugin === void 0 ? void 0 : plugin.id,
        });
        var cellOuterStyle = cellSpacing.y !== 0 || cellSpacing.x !== 0
            ? {
                padding: "".concat(cellSpacing.y / 2, "px ").concat(cellSpacing.x / 2, "px"),
            }
            : undefined;
        var innerStylingProps = (0, getCellStylingProps_1.getCellInnerDivStylingProps)(cell, plugin, data);
        return (react_1.default.createElement(Provider, __assign({}, props_1),
            react_1.default.createElement("div", { className: outerClasses, style: cellOuterStyle },
                react_1.default.createElement("div", __assign({}, innerStylingProps),
                    react_1.default.createElement("div", { style: hasInlineNeighbour
                            ? null
                            : {
                                display: 'flex',
                                flexDirection: 'column',
                                height: '100%',
                            } },
                        react_1.default.createElement(Renderer, __assign({}, props_1), ((_d = cell.rows) === null || _d === void 0 ? void 0 : _d.length) ? (react_1.default.createElement("div", { style: { margin: "".concat(-pluginCellSpacing_1.y / 2, "px 0") } }, (_e = cell.rows) === null || _e === void 0 ? void 0 : _e.map(function (r) { return (react_1.default.createElement(HTMLRow, __assign({ key: r.id }, r, { cellPlugins: childCellPlugins_1, cellSpacing: pluginCellSpacing_1, lang: lang }))); }))) : null))))));
    }
    else if (((_f = cell.rows) === null || _f === void 0 ? void 0 : _f.length) > 0) {
        return (react_1.default.createElement("div", { className: outerClasses, style: { padding: "0 ".concat(cellSpacing.x / 2, "px") } }, cell.rows.map(function (r) { return (react_1.default.createElement(HTMLRow, __assign({ key: r.id }, r, { lang: lang, cellPlugins: cellPlugins, cellSpacing: cellSpacing }))); })));
    }
    return (react_1.default.createElement("div", { className: outerClasses },
        react_1.default.createElement("div", { className: "react-page-cell-inner" })));
});
exports.HTMLRenderer = react_1.default.memo(function (_a) {
    var value = _a.value, cellPlugins = _a.cellPlugins, cellSpacing = _a.cellSpacing, _b = _a.lang, lang = _b === void 0 ? 'default' : _b;
    var data = (0, migrate_1.migrateValue)(value, { cellPlugins: cellPlugins, lang: lang });
    var normCellSpacing = (0, getCellSpacing_1.normalizeCellSpacing)(cellSpacing);
    if (!data) {
        return null;
    }
    var rows = data.rows;
    var optRows = (0, optimize_1.optimizeRows)(rows);
    return (react_1.default.createElement("div", { style: {
            margin: (optRows === null || optRows === void 0 ? void 0 : optRows.length) ? "".concat(-normCellSpacing.y / 2, "px 0") : null,
        } }, (0, setAllSizesAndOptimize_1.setAllSizesAndOptimize)(optRows).map(function (row) { return (react_1.default.createElement(HTMLRow, __assign({ key: row.id, cellPlugins: cellPlugins, lang: lang, cellSpacing: normCellSpacing }, row))); })));
});
