// @react-page/editor/lib/core/components/Cell/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var classnames_1 = __importDefault(require("classnames"));
var react_1 = __importStar(require("react"));
var getCellStylingProps_1 = require("../../utils/getCellStylingProps");
var hooks_1 = require("../hooks");
var ErrorCell_1 = __importDefault(require("./ErrorCell"));
var Handle_1 = __importDefault(require("./Handle"));
var Inner_1 = __importDefault(require("./Inner"));
var MoveActions_1 = __importDefault(require("./MoveActions"));
var scrollIntoViewWithOffset_1 = __importDefault(require("./utils/scrollIntoViewWithOffset"));
var CellErrorGate = /** @class */ (function (_super) {
    __extends(class_1, _super);
    function class_1() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            error: null,
        };
        return _this;
    }
    class_1.prototype.componentDidCatch = function (error) {
        this.setState({ error: error });
    };
    class_1.prototype.render = function () {
        if (this.state.error) {
            return react_1.default.createElement(ErrorCell_1.default, { nodeId: this.props.nodeId, error: this.state.error });
        }
        return this.props.children;
    };
    return class_1;
}(react_1.default.Component));
var Cell = function (_a) {
    var _b;
    var nodeId = _a.nodeId, measureRef = _a.measureRef;
    var focused = (0, hooks_1.useIsFocused)(nodeId);
    var _c = (0, hooks_1.useCellProps)(nodeId, function (_a) {
        var inline = _a.inline, hasInlineNeighbour = _a.hasInlineNeighbour, isDraft = _a.isDraft, isDraftI18n = _a.isDraftI18n, size = _a.size;
        return ({
            inline: inline,
            hasInlineNeighbour: hasInlineNeighbour,
            isDraft: isDraft,
            isDraftI18n: isDraftI18n,
            size: size,
        });
    }), inline = _c.inline, hasInlineNeighbour = _c.hasInlineNeighbour, isDraft = _c.isDraft, isDraftI18n = _c.isDraftI18n, size = _c.size;
    var lang = (0, hooks_1.useLang)();
    var isPreviewMode = (0, hooks_1.useIsPreviewMode)();
    var isResizeMode = (0, hooks_1.useIsResizeMode)();
    var isLayoutMode = (0, hooks_1.useIsLayoutMode)();
    var isInsertMode = (0, hooks_1.useIsInsertMode)();
    var multiNodesSelected = (0, hooks_1.useAllFocusedNodeIds)().length > 1;
    var hasChildren = (0, hooks_1.useNodeHasChildren)(nodeId);
    var showMoveButtons = (0, hooks_1.useOption)('showMoveButtonsInLayoutMode');
    var hasPlugin = (0, hooks_1.useCellHasPlugin)(nodeId);
    var _d = (0, hooks_1.useCellSpacing)(), cellSpacingX = _d.x, cellSpacingY = _d.y;
    var needVerticalPadding = !hasChildren || hasPlugin;
    var isDraftInLang = (_b = isDraftI18n === null || isDraftI18n === void 0 ? void 0 : isDraftI18n[lang]) !== null && _b !== void 0 ? _b : isDraft;
    var ref = react_1.default.useRef();
    var setReferenceNodeId = (0, hooks_1.useSetDisplayReferenceNodeId)();
    var onClick = (0, react_1.useCallback)(function (e) {
        if (isInsertMode) {
            e.stopPropagation();
            setReferenceNodeId(nodeId);
        }
    }, [nodeId, isInsertMode, setReferenceNodeId]);
    (0, hooks_1.useScrollToViewEffect)(nodeId, function () {
        if (ref.current)
            (0, scrollIntoViewWithOffset_1.default)(ref.current, 120); // 120 is just a sane default, we might make int configurable in the future
    }, [ref.current]);
    if (isDraftInLang && isPreviewMode) {
        return null;
    }
    var cellOuterStlye = cellSpacingY !== 0 || cellSpacingX !== 0
        ? {
            padding: "".concat(needVerticalPadding ? cellSpacingY / 2 : 0, "px ").concat(cellSpacingX / 2, "px"),
        }
        : undefined;
    return (react_1.default.createElement("div", { style: cellOuterStlye, ref: ref, className: (0, getCellStylingProps_1.getCellOuterDivClassName)({
            hasChildren: hasChildren,
            hasInlineNeighbour: hasInlineNeighbour,
            size: size,
            inline: inline,
        }) +
            ' ' +
            (0, classnames_1.default)({
                'react-page-cell-has-plugin': hasPlugin,
                'react-page-cell-focused': focused,
                'react-page-cell-is-draft': isDraftInLang,
                'react-page-cell-bring-to-front': !isResizeMode && !isLayoutMode && inline, // inline must not be active for resize/layout
            }), onClick: onClick },
        react_1.default.createElement(Handle_1.default, { nodeId: nodeId }),
        showMoveButtons &&
            isLayoutMode &&
            !hasChildren &&
            !multiNodesSelected ? (react_1.default.createElement(MoveActions_1.default, { nodeId: nodeId })) : null,
        react_1.default.createElement("div", { ref: measureRef, style: {
                height: '100%',
                boxSizing: 'border-box',
            } },
            react_1.default.createElement(CellErrorGate, { nodeId: nodeId },
                react_1.default.createElement(Inner_1.default, { nodeId: nodeId })))));
};
exports.default = react_1.default.memo(Cell);
