// @react-page/editor/lib/ui/Sidebar/UndoRedo/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Redo_1 = __importDefault(require("@material-ui/icons/Redo"));
var Undo_1 = __importDefault(require("@material-ui/icons/Undo"));
var react_1 = __importDefault(require("react"));
var hooks_1 = require("../../../core/components/hooks");
var index_1 = __importDefault(require("../Button/index"));
var UndoRedo = function (_a) {
    var labelUndo = _a.labelUndo, labelRedo = _a.labelRedo;
    var undo = (0, hooks_1.useUndo)();
    var canUndo = (0, hooks_1.useCanUndo)();
    var canRedo = (0, hooks_1.useCanRedo)();
    var redo = (0, hooks_1.useRedo)();
    var isSmall = (0, hooks_1.useIsSmallScreen)();
    return (react_1.default.createElement("div", { style: {
            height: isSmall ? 56 : 80,
            float: 'right',
            display: 'flex',
            direction: 'ltr',
            transform: 'scale(1.2)',
        } },
        react_1.default.createElement("div", { style: {
                width: isSmall ? 29 : 36,
                overflow: 'hidden',
                marginRight: isSmall ? 1 : 2,
            } },
            react_1.default.createElement(index_1.default, { active: true, disabled: !canUndo, style: {
                    transform: "translateX(".concat(isSmall ? 27 : 35, "px)"),
                }, icon: react_1.default.createElement(Undo_1.default, { style: { transform: "translateX(-".concat(isSmall ? 6 : 12, "px)") } }), description: labelUndo, onClick: undo, activeColor: "primary" })),
        react_1.default.createElement("div", { style: {
                width: isSmall ? 28 : 36,
                overflow: 'hidden',
                marginLeft: 1,
            } },
            react_1.default.createElement(index_1.default, { style: {
                    position: 'relative',
                    transform: 'translateX(1px)',
                }, active: true, disabled: !canRedo, icon: react_1.default.createElement(Redo_1.default, { style: { transform: "translateX(".concat(isSmall ? 6 : 12, "px)") } }), description: labelRedo, onClick: redo, activeColor: "primary" }))));
};
exports.default = react_1.default.memo(UndoRedo);
