// @react-page/editor/lib/core/reducer/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initialState = exports.reducer = void 0;
var redux_1 = require("redux");
var values_1 = require("./values");
var display_1 = require("./display");
var focus_1 = require("./focus");
var hover_1 = require("./hover");
var settings_1 = require("./settings");
var reducer = (0, redux_1.combineReducers)({
    values: values_1.values,
    display: display_1.display,
    focus: focus_1.focus,
    settings: settings_1.settings,
    hover: hover_1.hover,
    __nodeCache: function () { return null; }, // always empty __nodeCache
});
exports.reducer = reducer;
exports.default = (0, redux_1.combineReducers)({ reactPage: reducer });
function initialState(value, lang) {
    return {
        reactPage: {
            __nodeCache: {},
            hover: null,
            focus: null,
            display: {
                mode: 'edit',
                zoom: 1,
            },
            settings: {
                lang: lang,
            },
            values: {
                past: [],
                present: value,
                future: [],
            },
        },
    };
}
exports.initialState = initialState;
