// @react-page/editor/lib/core/components/Editable/Inner/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var hooks_1 = require("../../hooks");
var useKeepScrollPosition_1 = require("../hooks/useKeepScrollPosition");
var Rows_1 = __importDefault(require("./Rows"));
var Inner = function () {
    var _a;
    var mode = (0, hooks_1.useDisplayMode)();
    var ref = (0, react_1.useRef)();
    var zoom = (0, hooks_1.useZoom)();
    (0, useKeepScrollPosition_1.useKeepScrollPosition)(ref);
    var rect = (_a = ref.current) === null || _a === void 0 ? void 0 : _a.getBoundingClientRect();
    var zoomTransformOriginY = window.innerHeight / 2 - (rect === null || rect === void 0 ? void 0 : rect.top);
    var offsetPercent = zoomTransformOriginY / (rect === null || rect === void 0 ? void 0 : rect.height);
    var backdropPercent = 50 * (1 - zoom);
    var left = backdropPercent + '%';
    var right = 100 - backdropPercent + '%';
    var top = backdropPercent * offsetPercent * 2 + '%';
    var bottom = 100 - backdropPercent * (1 - offsetPercent) * 2 + '%';
    return (react_1.default.createElement("div", { ref: ref, style: {
            position: 'relative',
        } },
        react_1.default.createElement("div", { style: {
                opacity: zoom < 1 ? 1 : 0,
                position: 'absolute',
                top: 0,
                left: 0,
                right: 0,
                bottom: 0,
                transition: '0.6s',
                clipPath: "polygon(0% 0%, 0% 100%, ".concat(left, " 100%, ").concat(left, " ").concat(top, ", ").concat(right, " ").concat(top, ", ").concat(right, " ").concat(bottom, ", ").concat(left, " ").concat(bottom, ", ").concat(left, " 100%, 100% 100%, 100% 0%)"),
                background: "url(\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='16' height='16  ' viewBox='0 0 8 8'%3E%3Cg fill='%23c5c5c5' fill-opacity='0.4'%3E%3Cpath fill-rule='evenodd' d='M0 0h4v4H0V0zm4 4h4v4H4V4z'/%3E%3C/g%3E%3C/svg%3E\")",
            } }),
        react_1.default.createElement("div", { style: {
                transformOrigin: "center ".concat(zoomTransformOriginY, "px"),
                transform: "scale(".concat(zoom, ")"),
                transition: '0.6s',
            }, className: 'react-page-editable react-page-editable-mode-' + mode },
            react_1.default.createElement(Rows_1.default, null))));
};
exports.default = react_1.default.memo(Inner);
