// @react-page/editor/lib-es/ui/PluginDrawer/index.js
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
import Drawer from '@material-ui/core/Drawer';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListSubheader from '@material-ui/core/ListSubheader';
import TextField from '@material-ui/core/TextField';
import React from 'react';
import { Portal } from 'react-portal';
import { useIsInsertMode, useUiTranslator, useDisplayModeReferenceNodeId, useAllCellPluginsForNode, } from '../../core/components/hooks';
import Item from './Item/index';
var getPluginTitle = function (plugin) { var _a; return (_a = (plugin.title || plugin.text)) !== null && _a !== void 0 ? _a : ''; };
export var PluginDrawer = React.memo(function () {
    var defaultLabels = {
        noPluginFoundContent: 'No blocks found',
        searchPlaceholder: 'Search for blocks',
        insertPlugin: 'Add blocks to page',
        dragMe: 'Drag me!',
    };
    var nodeId = useDisplayModeReferenceNodeId();
    var plugins = useAllCellPluginsForNode(nodeId);
    var t = useUiTranslator().t;
    var _a = __read(React.useState(''), 2), searchText = _a[0], setSearchText = _a[1];
    var searchFilter = React.useCallback(function (plugin) {
        var id = plugin.id;
        var title = getPluginTitle(plugin);
        return (plugin &&
            id &&
            !plugin.hideInMenu &&
            (id.toLowerCase().startsWith(searchText === null || searchText === void 0 ? void 0 : searchText.toLowerCase()) ||
                (plugin.description &&
                    plugin.description
                        .toLowerCase()
                        .startsWith(searchText === null || searchText === void 0 ? void 0 : searchText.toLowerCase())) ||
                (title && title.toLowerCase().startsWith(searchText === null || searchText === void 0 ? void 0 : searchText.toLowerCase()))));
    }, [searchText]);
    var onSearch = React.useCallback(function (e) {
        var target = e.target;
        if (target instanceof HTMLInputElement) {
            setSearchText(target.value);
        }
    }, [setSearchText]);
    var isInsertMode = useIsInsertMode();
    var inputRef = React.useRef();
    React.useEffect(function () {
        var handle;
        if (inputRef.current && isInsertMode) {
            handle = setTimeout(function () {
                var e = inputRef.current.querySelector('input');
                if (e) {
                    e.focus();
                }
            }, 100);
        }
        return function () {
            clearTimeout(handle);
        };
    }, [inputRef.current, isInsertMode]);
    var filteredPlugins = plugins.filter(searchFilter);
    return (React.createElement(Portal, null,
        React.createElement(Drawer, { variant: "persistent", className: "react-page-plugin-drawer", open: isInsertMode, PaperProps: {
                style: {
                    width: 320,
                },
            } },
            React.createElement(List, { subheader: React.createElement(ListSubheader, null, t(defaultLabels.insertPlugin)) },
                React.createElement(ListItem, null,
                    React.createElement(TextField, { inputRef: inputRef, placeholder: t(defaultLabels.searchPlaceholder), fullWidth: true, onChange: onSearch })),
                filteredPlugins.length === 0 && (React.createElement(ListSubheader, null, t(defaultLabels.noPluginFoundContent)))),
            filteredPlugins.length > 0 && (React.createElement(List, null, filteredPlugins.map(function (plugin, k) {
                return (React.createElement(Item, { translations: defaultLabels, plugin: plugin, key: k.toString(), insert: {
                        plugin: plugin.id,
                    } }));
            }))))));
});
