// @react-page/editor/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VERSION = exports.deepEquals = exports.migrateValue = exports.getTextContents = exports.createValue = exports.makeUniformsSchema = exports.Migration = exports.lazyLoad = exports.objIsNode = void 0;
//import './wdyr';
__exportStar(require("./core/types"), exports);
__exportStar(require("./core/components/hooks"), exports);
__exportStar(require("./ui"), exports);
var lazyLoad_1 = __importDefault(require("./core/helper/lazyLoad"));
exports.lazyLoad = lazyLoad_1.default;
var Migration_1 = require("./core/migrations/Migration");
Object.defineProperty(exports, "Migration", { enumerable: true, get: function () { return Migration_1.Migration; } });
var Editor_1 = __importDefault(require("./editor/Editor"));
var makeUniformsSchema_1 = __importDefault(require("./ui/AutoformControls/makeUniformsSchema"));
exports.makeUniformsSchema = makeUniformsSchema_1.default;
var migrate_1 = require("./core/migrations/migrate");
Object.defineProperty(exports, "migrateValue", { enumerable: true, get: function () { return migrate_1.migrateValue; } });
var deepEquals_1 = __importDefault(require("./core/utils/deepEquals"));
exports.deepEquals = deepEquals_1.default;
var createValue_1 = require("./core/utils/createValue");
Object.defineProperty(exports, "createValue", { enumerable: true, get: function () { return createValue_1.createValue; } });
var objIsNode_1 = require("./core/utils/objIsNode");
Object.defineProperty(exports, "objIsNode", { enumerable: true, get: function () { return objIsNode_1.objIsNode; } });
var getTextContents_1 = require("./core/utils/getTextContents");
Object.defineProperty(exports, "getTextContents", { enumerable: true, get: function () { return getTextContents_1.getTextContents; } });
exports.default = Editor_1.default;
exports.VERSION = '4.9.0';
