// @react-page/editor/lib/ui/BottomToolbar/Drawer.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BottomToolbarDrawer = void 0;
var core_1 = require("@material-ui/core");
var react_1 = __importStar(require("react"));
var hooks_1 = require("../../core/components/hooks");
var darkBlack = 'rgba(0, 0, 0, 0.87)';
var bright = 'rgba(255,255,255, 0.98)';
var brightBorder = 'rgba(0, 0, 0, 0.12)';
var BottomToolbarDrawer = function (_a) {
    var className = _a.className, anchor = _a.anchor, open = _a.open, dark = _a.dark, _b = _a.scale, scale = _b === void 0 ? 1 : _b, children = _a.children, _c = _a.style, style = _c === void 0 ? {} : _c;
    var divider = (react_1.default.createElement(core_1.Divider, { style: {
            marginLeft: -24,
            marginRight: -24,
            marginTop: 12,
            marginBottom: 12,
        } }));
    var theChildren = react_1.default.Children.toArray(children).filter(Boolean);
    var isSmall = (0, hooks_1.useIsSmallScreen)();
    return (react_1.default.createElement(core_1.Portal, null,
        react_1.default.createElement(core_1.Drawer, { SlideProps: {
                mountOnEnter: true,
                unmountOnExit: true,
            }, variant: "persistent", className: className, open: open, anchor: anchor, PaperProps: {
                style: {
                    zIndex: 10,
                    backgroundColor: 'transparent',
                    border: 'none',
                    overflow: 'visible',
                    pointerEvents: 'none',
                },
            } },
            react_1.default.createElement("div", { style: __assign(__assign(__assign({ pointerEvents: 'all', border: "".concat(dark ? darkBlack : brightBorder, " 1px solid"), borderRadius: '4px 4px 0 0', backgroundColor: dark ? darkBlack : bright, padding: '12px 24px' }, (isSmall
                    ? {
                        marginLeft: 20,
                        marginRight: 80,
                    }
                    : {
                        marginLeft: 'auto',
                        marginRight: 'auto',
                        minWidth: '50vw',
                        maxWidth: 'min(1280px, calc(100vw - 250px))',
                    })), { boxShadow: '0px 1px 8px -1px rgba(0,0,0,0.4)', position: 'relative', transformOrigin: 'bottom', transform: "scale(".concat(scale, ")"), transition: 'scale 0.3s' }), style) }, theChildren.map(function (child, index) { return (react_1.default.createElement(react_1.Fragment, { key: index },
                child,
                index < theChildren.length - 1 ? divider : null)); })))));
};
exports.BottomToolbarDrawer = BottomToolbarDrawer;
