// @react-page/editor/lib/core/components/Cell/PluginComponent/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var ui_1 = require("../../../../ui");
var hooks_1 = require("../../hooks");
var PluginControls_1 = __importDefault(require("../PluginControls"));
var PluginMissing_1 = __importDefault(require("../PluginMissing"));
var NoopProvider_1 = __importDefault(require("../NoopProvider"));
var PluginComponent = function (_a) {
    var _b, _c, _d;
    var nodeId = _a.nodeId, children = _a.children, hasChildren = _a.hasChildren;
    var lang = (0, hooks_1.useLang)();
    var components = (0, hooks_1.useOption)('components');
    var CustomPluginMissing = components === null || components === void 0 ? void 0 : components.CellPluginMissing;
    var isPreviewMode = (0, hooks_1.useIsPreviewMode)();
    var isEditMode = (0, hooks_1.useIsEditMode)();
    var _e = __read((0, hooks_1.useDebouncedCellData)(nodeId), 2), data = _e[0], onChange = _e[1];
    var pluginId = (0, hooks_1.useCellProps)(nodeId, function (c) { var _a; return (_a = c.plugin) === null || _a === void 0 ? void 0 : _a.id; });
    var plugin = (0, hooks_1.usePluginOfCell)(nodeId);
    var focused = (0, hooks_1.useIsExclusivlyFocused)(nodeId);
    var hasInlineNeighbour = (0, hooks_1.useCellProps)(nodeId, function (c) { return c.hasInlineNeighbour; });
    var Renderer = plugin === null || plugin === void 0 ? void 0 : plugin.Renderer;
    var Missing = CustomPluginMissing !== null && CustomPluginMissing !== void 0 ? CustomPluginMissing : PluginMissing_1.default;
    var Provider = (_b = plugin === null || plugin === void 0 ? void 0 : plugin.Provider) !== null && _b !== void 0 ? _b : NoopProvider_1.default;
    var remove = (0, hooks_1.useRemoveCell)(nodeId);
    var Toolbar = (_c = components === null || components === void 0 ? void 0 : components.BottomToolbar) !== null && _c !== void 0 ? _c : ui_1.BottomToolbar;
    var componentProps = (0, react_1.useMemo)(function () { return ({
        nodeId: nodeId,
        lang: lang,
        data: data,
        pluginConfig: plugin,
        focused: isEditMode && focused,
        readOnly: !isEditMode,
        onChange: onChange,
        isEditMode: isEditMode,
        isPreviewMode: isPreviewMode,
        remove: remove,
    }); }, [
        nodeId,
        lang,
        data,
        plugin,
        isEditMode,
        focused,
        onChange,
        isEditMode,
        isPreviewMode,
        remove,
    ]);
    // In case of non-zero cell spacing, nested layouts (layout plugins with children) should have their
    // margin collapsing functionality off. The simplest solution is to use display:flex for the below wrapping <div>.
    // This however is not compatible with inline elements flotation, so if a cell has inline neighbors, we are going
    // to have to keep display:block style. Layout plugins with inline cell support will have to take care of
    // margin collapsing internally on their own.
    var display = hasInlineNeighbour
        ? {}
        : {
            display: 'flex',
            flexDirection: 'column',
        };
    return (react_1.default.createElement(Provider, __assign({}, componentProps),
        react_1.default.createElement(react_1.default.Fragment, null,
            react_1.default.createElement("div", { style: __assign(__assign({}, display), { height: '100%', pointerEvents: !isPreviewMode &&
                        plugin &&
                        !(plugin === null || plugin === void 0 ? void 0 : plugin.allowClickInside) &&
                        !hasChildren
                        ? 'none'
                        : undefined }) }, Renderer ? (react_1.default.createElement(Renderer, __assign({}, componentProps), children)) : pluginId ? (react_1.default.createElement(Missing, __assign({}, componentProps, { pluginId: pluginId }))) : (children)),
            react_1.default.createElement(Toolbar, { nodeId: nodeId, open: focused, dark: (_d = plugin === null || plugin === void 0 ? void 0 : plugin.bottomToolbar) === null || _d === void 0 ? void 0 : _d.dark, pluginControls: isEditMode && (plugin === null || plugin === void 0 ? void 0 : plugin.controls) ? (react_1.default.createElement(PluginControls_1.default, { componentProps: componentProps, controls: plugin === null || plugin === void 0 ? void 0 : plugin.controls })) : null }))));
};
exports.default = PluginComponent;
