// @react-page/editor/lib/core/defaultOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_RENDER_OPTIONS = exports.DEFAULT_OPTIONS = void 0;
var react_dnd_html5_backend_1 = require("react-dnd-html5-backend");
var display_1 = require("./actions/display");
exports.DEFAULT_OPTIONS = {
    allowMoveInEditMode: true,
    allowResizeInEditMode: true,
    childConstraints: null,
    components: {},
    languages: null,
    uiTranslator: null,
    zoomEnabled: true,
    zoomFactors: [1, 0.75, 0.5, 0.25],
    undoRedoEnabled: true,
    editEnabled: true,
    insertEnabled: true,
    layoutEnabled: true,
    resizeEnabled: true,
    previewEnabled: true,
    dndBackend: react_dnd_html5_backend_1.HTML5Backend,
    blurGateDefaultMode: display_1.DISPLAY_MODE_EDIT,
    blurGateDisabled: false,
    middleware: [],
    store: null,
    hideEditorSidebar: false,
    showMoveButtonsInBottomToolbar: true,
    showMoveButtonsInLayoutMode: true,
    sidebarPosition: 'rightAbsolute',
    customOptions: [],
};
exports.DEFAULT_RENDER_OPTIONS = {
    cellPlugins: [],
    cellSpacing: null,
};
