// @react-page/editor/lib/core/EditorStore.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEmptyState = exports.EditorContext = void 0;
var store_1 = __importDefault(require("./store"));
var setting_1 = require("./actions/setting");
var editable_1 = require("./selector/editable");
var react_1 = require("react");
var createId_1 = require("./utils/createId");
exports.EditorContext = (0, react_1.createContext)(null);
var EditorStore = /** @class */ (function () {
    function EditorStore(_a) {
        var _b = _a.middleware, middleware = _b === void 0 ? [] : _b, store = _a.store, initialState = _a.initialState;
        var _this = this;
        this.getNodeWithAncestors = function (nodeId) {
            return (0, editable_1.findNodeInState)(_this.store.getState(), nodeId);
        };
        this.getNode = function (nodeId) {
            var _a;
            return (_a = (0, editable_1.findNodeInState)(_this.store.getState(), nodeId)) === null || _a === void 0 ? void 0 : _a.node;
        };
        this.store = store || (0, store_1.default)(initialState, middleware);
        this.middleware = middleware;
    }
    EditorStore.prototype.setLang = function (lang) {
        this.store.dispatch((0, setting_1.setLang)(lang));
    };
    return EditorStore;
}());
var createEmptyState = function () {
    return ({ id: (0, createId_1.createId)(), rows: [] });
};
exports.createEmptyState = createEmptyState;
exports.default = EditorStore;
