// @react-page/editor/lib-es/ui/Sidebar/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import React from 'react';
import { useOption, useUiTranslator } from '../../core/components/hooks';
import ToggleEdit from './ToggleEdit/index';
import ToggleInsert from './ToggleInsert/index';
import ToggleLayout from './ToggleLayout/index';
import TogglePreview from './TogglePreview/index';
import ToggleResize from './ToggleResize/index';
import UndoRedo from './UndoRedo';
import Zoom from './Zoom';
var getStickyNessstyle = function (stickyness) {
    if (!stickyness ||
        (!stickyness.shouldStickToBottom && !stickyness.shouldStickToTop)) {
        return {
            position: 'fixed',
            right: stickyness.rightOffsetFixed || 0,
        };
    }
    return {
        position: 'absolute',
        bottom: stickyness.shouldStickToBottom ? 0 : 'auto',
        top: stickyness.shouldStickToTop ? 0 : 'auto',
        right: -stickyness.rightOffset || 0,
    };
};
export var Sidebar = function (_a) {
    var stickyNess = _a.stickyNess;
    var t = useUiTranslator().t;
    var zoomEnabled = useOption('zoomEnabled');
    var undoRedoEnabled = useOption('undoRedoEnabled');
    var editEnabled = useOption('editEnabled');
    var insertEnabled = useOption('insertEnabled');
    var layoutEnabled = useOption('layoutEnabled');
    var resizeEnabled = useOption('resizeEnabled');
    var previewEnabled = useOption('previewEnabled');
    var defaultLabels = {
        edit: 'Edit blocks',
        insert: 'Add blocks',
        layout: 'Move blocks',
        resize: 'Resize blocks',
        preview: 'Preview page',
    };
    var customOptions = useOption('customOptions');
    var actions = __spreadArray([
        // eslint-disable-next-line react/jsx-key
        undoRedoEnabled
            ? { action: React.createElement(UndoRedo, { labelRedo: "redo", labelUndo: "undo" }) }
            : null,
        zoomEnabled
            ? { action: React.createElement(Zoom, { labelZoomIn: "zoom in", labelZoomOut: "zoom out" }) }
            : null,
        editEnabled
            ? { action: React.createElement(ToggleEdit, { label: t(defaultLabels.edit) }) }
            : null,
        insertEnabled
            ? { action: React.createElement(ToggleInsert, { label: t(defaultLabels.insert) }) }
            : null,
        layoutEnabled
            ? { action: React.createElement(ToggleLayout, { label: t(defaultLabels.layout) }) }
            : null,
        resizeEnabled
            ? { action: React.createElement(ToggleResize, { label: t(defaultLabels.resize) }) }
            : null,
        previewEnabled
            ? { action: React.createElement(TogglePreview, { label: t(defaultLabels.preview) }) }
            : null
    ], __read(customOptions.map(function (customOption) { return ({ action: customOption }); })), false).filter(Boolean);
    return (React.createElement("div", { className: "react-page-controls-mode-toggle-control-group", style: __assign({ position: 'fixed', zIndex: 10001, bottom: 0, right: 0, display: 'flex', maxHeight: '100%' }, getStickyNessstyle(stickyNess)) },
        React.createElement("div", { ref: stickyNess.stickyElRef, style: {
                padding: 16,
                position: 'relative',
                flexFlow: 'column wrap',
                direction: 'rtl',
                display: 'flex',
            } }, actions.map(function (_a, index) {
            var action = _a.action;
            return (React.createElement("div", { key: index, className: "react-page-controls-mode-toggle-control", style: {
                    animationDelay: (actions.length - index) * 150 + 'ms',
                } },
                action,
                React.createElement("div", { className: "react-page-controls-mode-toggle-clearfix" })));
        }))));
};
